### ISSUE TITLE
Implicit Unit Test Ordering

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#200]
Check for any implicit unit test ordering requirements across all our tests (there is a tool to do that, just forget the name atm).

If found, make the ordering explicit in code so that it is enforced (and transparently obvious) or refactor the test.

**Comment by janden** [comment]
Out of curiosity, what would this mean in practice? That we would test lower-level components first to avoid spending time on longer tests that are more high-level?

**Comment by garrettwrong** [comment]
No... we don't want hidden implicit ordering, that is bad.

We want tests to be atomic and idempotent _or_ have very clear explicit ordering.

If a test has implicit ordering and is mutating state, the collection of tests may only be working by happenstance. Bad times.

When building up a collection of data structures in test cases for use later in a more sophisticated test case, we want to spell out/enforce that happens in a sane order. It should be clear that Test B depends on Test A existing and succeeding.  This was something that came up in a recent PR discussion.  I'm sure you'll find that thread as you review, maybe for context, but generally I think not relying on such behavior is a pretty general practice.

**Comment by janden** [comment]
Ah I see now. I was assuming that all tests were already atomic and idempotent, but it's good to check that.

**Comment by vineetbansal** [comment]
I've verified that there is no assumed implicit ordering of tests in our code (by inspecting code as well as using the `pytest-random-order` package). Maybe the steps we take are:

- Enable `dev` section tools to be installed by default in travis/azure.
- Add `pytest-random-order` to this `dev` section (which seems more popular than it's 1 alternative).
- Use the following options in `pytest.ini`:
   ```
    addopts = -m "not expensive" --random-order
    ```

If this looks good I can work with @junchaoxia  to open up a PR.

In general one has to bend over backwards to write tests that *do* depend on each other (by messing around with `cls`). Using `@classmethod` is one way to do it. I'm doing this at one place in `test_covar3d`, for performance reasons. But since this is a `setUpClass` and not `setUp`, this is only entered once, and not repeatedly on every test method.

I would be wary of `@classmethod` being used with `setUp`, which does open up this possibility of maintaining state between test methods, like the code we saw yesterday.

**Comment by garrettwrong** [comment]
Yes, I'm glad there is not any cases in the code today... that's great. At the time this issue was open, we were definitely trying (unsuccsessfully) to do such things. We ended up stuffing all that code into `setUp`... the one that was a classmethod ... That code has been screenshared in several dev meetings over the last couple months. The discussions were not summarized here.

Sounds good to me.

**Comment by junchaoxia** [comment]
I am trying to make sure that we are in the same page. Currently we have two kinds of unit tests in general. The first one is that each unit test in the same TestCase class is independent to each other. The second one is that a unit test could depend on the evaluated variables in previous unit tests unless the previous calculations are repeated again within the same unit test.  For the later case (such as test_orient_sync.py) we have used the `setup` to run all calculations first and evaluated them one by one in each unit test for the selected results.  For both cases, we did not imply the order and they can be run with random.  But we can change the setting as  @vineetbansal suggested to run the tests in random order explicitly instead of the alphabet order in default.
